x-pack/filebeat/input/o365audit/listblobs.go

```diff
@@ -0,0 +1,296 @@
+		}
+		if entry.Created.Before(l.startTime) {
+			l.env.Logger.Errorf("! Event created before query")
+		}
+		if entry.Created.After(l.endTime) {
+			l.env.Logger.Errorf("! Event created after query")
+		}
+	}
+	// Fetch the next page if a NextPageUri header is found.
+	if url, found := getNextPage(response); found {
+		return append(actions, poll.Fetch(newPager(url, l)))
+	}
+	// Otherwise fetch the next time window.
+	return append(actions, poll.Fetch(l.Next()))
+}
+
+// Next returns a listBlob that will fetch events in future.
+func (l listBlob) Next() listBlob {
+	return l.adjustTimes(l.endTime)
+}
+
+var fatalErrors = map[string]struct{}{
+	// Missing parameter: {0}.
+	"AF20001": {},
+	// Invalid parameter type: {0}. Expected type: {1}
+	"AF20002": {},
+	// Expiration {0} provided is set to past date and time.
+	"AF20003": {},
+	// The tenant ID passed in the URL ({0}) does not match the tenant ID passed in the access token ({1}).
+	"AF20010": {},
+	// Specified tenant ID ({0}) does not exist in the system or has been deleted.
+	"AF20011": {},
+	// Specified tenant ID ({0}) is incorrectly configured in the system.
+	"AF20012": {},
+	// The tenant ID passed in the URL ({0}) is not a valid GUID.
+	"AF20013": {},
+	// The specified content type is not valid.
+	"AF20020": {},
+	// The webhook endpoint {{0}) could not be validated. {1}
+	"AF20021": {},
+}
+
+func (l listBlob) handleError(response *http.Response) (actions []poll.Action) {
+	var msg apiError
+	readJSONBody(response, &msg)
+	l.env.Logger.Warnf("Got error %s: %+v", response.Status, msg)
+	l.delay = l.env.Config.ErrorRetryInterval
+
+	switch response.StatusCode {
+	case 401:
+		// Authentication error. Renew oauth token and repeat this op.
+		l.delay = l.env.Config.PollInterval
+		return []poll.Action{
+			poll.RenewToken(),
+			poll.Fetch(l),
+		}
+	case 408, 503:
+		// Known errors when the backend is down.
+		//Repeat the request without reporting an error.
+		return []poll.Action{
+			poll.Fetch(l),
+		}
+	}
+
+	if _, found := fatalErrors[msg.Error.Code]; found {
+		return []poll.Action{
+			l.env.ReportAPIError(msg),
+			poll.Terminate(errors.New(msg.Error.Message)),
+		}
+	}
+
+	switch msg.Error.Code {
+	// AF20022: No subscription found for the specified content type
+	// AF20023: The subscription was disabled by [..]
+	case "AF20022", "AF20023":
+		l.delay = 0
+		// Subscribe and retry
+		return []poll.Action{
+			poll.Fetch(Subscribe(l.env)),
+			poll.Fetch(l),
+		}
+	// AF20030: Start time and end time must both be specified (or both omitted) and must
+	// be less than or equal to 24 hours apart, with the start time no more than
+	// 7 days in the past.
+	// AF20055: (Same).
+	case "AF20030", "AF20055":
+		// As of writing this, the server fails a request if it's more than
+		// retention_time(7d)+1h in the past.
+		// On the other hand, requests can be days into the future without error.
+
+		// First check if this is caused by a request close to  that's been
```

Comment:
Is there a missing word here?
```suggestion
		// First check if this is caused by a request close to the max retention period that's been
```